client/with_dry_runs.go

```diff
@@ -0,0 +1,236 @@
+	// If the gas limit is set to 0, ethereum client will do the estimation for us.
+	// We only force the estimation if the gas limit is set to a non zero value.
+	if glp.GetGasLimit() == 0 {
+		return 0, nil
+	}
+	parsed, err := abi.JSON(strings.NewReader(binding))
+	if err != nil {
+		return 0, errors.Wrap(err, "could not deserialize ABI")
+	}
+	input, err := parsed.Pack(method, params...)
+	if err != nil {
+		return 0, errors.Wrap(err, "could not pack input")
+	}
+	msg := ethereum.CallMsg{From: from, To: &to, Data: input}
+	gas, err := cwdr.bc.EstimateGas(msg)
+	return gas, errors.Wrap(err, "could not estimate gas")
+}
+
+// TransferMyst transfers myst
+func (cwdr *WithDryRuns) TransferMyst(req TransferRequest) (tx *types.Transaction, err error) {
+	_, err = cwdr.dryRun(
+		req,
+		bindings.MystTokenABI,
+		req.Identity,
+		req.MystAddress,
+		"transfer",
+		req.Recipient,
+		req.Amount,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "transaction dry run failed")
+	}
+
+	return cwdr.bc.TransferMyst(req)
+}
+
+// SettlePromise is settling the given consumer issued promise
+func (cwdr *WithDryRuns) SettlePromise(req SettleRequest) (*types.Transaction, error) {
+	rBytes := [32]byte{}
+	copy(rBytes[:], req.Promise.R)
+	lock := rBytes
+
+	_, err := cwdr.dryRun(
+		req,
+		bindings.ChannelImplementationABI,
+		req.Identity,
+		req.ChannelID,
+		"settlePromise",
+		big.NewInt(0).SetUint64(req.Promise.Amount),
+		big.NewInt(0).SetUint64(req.Promise.Fee),
+		lock,
+		req.Promise.Signature,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "transaction dry run failed")
+	}
+
+	return cwdr.SettlePromise(req)
+}
+
+// RegisterIdentity registers the given identity on blockchain
+func (cwdr *WithDryRuns) RegisterIdentity(rr RegistrationRequest) (*types.Transaction, error) {
+	_, err := cwdr.dryRun(
+		rr,
+		bindings.RegistryABI,
+		rr.Identity,
+		rr.RegistryAddress,
+		"registerIdentity",
+		rr.AccountantID,
+		rr.Loan,
+		rr.TransactorFee,
+		rr.Beneficiary,
+		rr.Signature,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "transaction dry run failed")
+	}
+
+	return cwdr.bc.RegisterIdentity(rr)
+}
+
+// SettleAndRebalance is settling given accountant issued promise
+func (cwdr *WithDryRuns) SettleAndRebalance(req SettleAndRebalanceRequest) (*types.Transaction, error) {
+	_, err := cwdr.dryRun(
+		req,
+		bindings.AccountantImplementationABI,
+		req.Identity,
+		req.AccountantID,
+		"settleAndRebalance",
+		toBytes32(req.Promise.ChannelID),
+		big.NewInt(0).SetUint64(req.Promise.Amount),
+		big.NewInt(0).SetUint64(req.Promise.Fee),
+		toBytes32(req.Promise.R),
+		req.Promise.Signature,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "transaction dry run failed")
+	}
+	return cwdr.bc.SettleAndRebalance(req)
+}
+
+// GetAccountantFee fetches the accountant fee from blockchain
+func (cwdr *WithDryRuns) GetAccountantFee(accountantAddress common.Address) (uint16, error) {
+	return cwdr.bc.GetAccountantFee(accountantAddress)
```

Comment:
It's strange to use WithDryRuns which does it only for some of the methods. Maybe better add dry run logic to whose few SettleAndRebalance, RegisterIdentity, SettlePromise, TransferMyst.

Reply:
I wanted to be able to use the same interface everywhere but still having the posibility to use the client both with and without retries, or with or without the dry run. The current design allows the end-user to combine the clients in any way shape of form, allowing for dry runs which will then retry the actual transaction.